File: src/main/java/org/example/SecureUtilities.java
package org.example;

import at.favre.lib.crypto.bcrypt.BCrypt;
import kotlin.Pair;

import java.security.SecureRandom;
import java.util.Random;

public class SecureUtilities {
    private static final Random random = new SecureRandom();
    private static final BCrypt.Hasher hasher = BCrypt.withDefaults();
    private static final BCrypt.Verifyer verifier = BCrypt.verifyer();
    private static final Config config = Config.getConfig();
    public static String generateSalt(int length) {
        StringBuilder salt = new StringBuilder();
        for (int i = 0; i < length; i++) {
            salt.append((char) (random.nextInt(26) + 'a'));
        }
        return salt.toString();
    }
    public static Pair<String, String> hash(String password) {
        String salt = generateSalt(config.saltSize);
        return new Pair<>(hasher.hashToString(config.hashCost, (salt + password).toCharArray()), salt);
    }
    public static boolean verifyHash(String password, String salt, String hashedPassword) {
        return verifier.verify((salt + password).toCharArray(), hashedPassword).verified;
    }
}

File: src/main/kotlin/SecureUtilities.kt
package org.example

import at.favre.lib.crypto.bcrypt.BCrypt
import java.security.SecureRandom
import java.util.*

private val random: Random = SecureRandom()
private val hasher: BCrypt.Hasher = BCrypt.withDefaults()
private val verifier: BCrypt.Verifyer = BCrypt.verifyer()
private val config: Config = Config.getConfig()
data class SaltHash(val salt: String, val hash: String)
fun generateSalt(length: Int): String {
    val salt: StringBuilder = StringBuilder()
    for (i in 0 until length) {
        salt.append((random.nextInt(26) + 'a'.code).toChar())
    }
    return salt.toString()
}
fun hashPassword(password: String): SaltHash {
    val salt: String = generateSalt(config.securitySettings.saltSize)
    val hash: String = hasher.hashToString(config.securitySettings.hashCost, (salt + password).toCharArray())
    return SaltHash(salt, hash)
}
fun verifyPassword(password: String, saltHash: SaltHash): Boolean {
    return verifier.verify((saltHash.salt + password).toCharArray(), saltHash.hash).verified
}